lib/logstash/inputs/http_poller/state_handler.rb

```diff
@@ -0,0 +1,133 @@
+# encoding: utf-8
+require "logstash/namespace"
+require "fileutils"
+
+module LogStash
+  module Inputs
+    module HTTPPoller
+      class StateHandler
+        def initialize(logger, requests)
+          @logger = logger
+          @state_path = ::File.join(LogStash::SETTINGS.get_value("path.data"), "plugins", "inputs", "http_poller", "state")
+          FileUtils.mkpath @state_path
+          @requests = requests
+          @pages_mutex = Mutex.new
+          @in_progress_pages = nil
+          @pages_signal = ConditionVariable.new
+          @stop_writer = false
+          @state_writer_thread = nil
+        end
+
+        attr_reader :in_progress_pages
+
+        public
+        def signal_waiting_threads
+          @pages_signal.broadcast
+        end
+
+        private
+        def get_state_file_path(name)
+          return ::File.join(@state_path, "state_" + name)
+        end
+
+        private
+        def atomic_state_write(file_path, current_page)
+          ::File.write(file_path + ".tmp", YAML.dump([@in_progress_pages.to_a, current_page.get]), mode: "w")
```

Comment:
This yaml file will end up looking something like the following which isn't very readable:

~~~ yaml
- - 7
  - 9
  - 11
- 13
~~~

I would prefer a file whose contents are self-descriptive:

~~~ yaml
---
in_progress_pages:
- 7
- 9
- 11
current_page: 12
~~~

```suggestion
          ::File.write(file_path + ".tmp", YAML.dump({"in_progress_pages" => @in_progress_pages.to_a, "current_page" => current_page.get}), mode: "w")
```